File: app/src/main/java/slak/gitforandroid/SelectableAdapterModel.java
package slak.gitforandroid;

import java.util.ArrayList;

public class SelectableAdapterModel<T> {
  private T thing;
  private boolean selectStatus = false;

  public SelectableAdapterModel(T thing) {
    this.thing = thing;
  }

  public void setSelectStatus(boolean newStatus) {
    selectStatus = newStatus;
  }
  public boolean isSelected() {
    return selectStatus;
  }
  public T getThing() {
    return thing;
  }

  public static <T> ArrayList<Integer> getSelectedModels(ArrayList<SelectableAdapterModel<T>> models) {
    ArrayList<Integer> selectedIndices = new ArrayList<>();
    for (int i = 0; i < models.size(); i++)
      if (models.get(i).selectStatus) selectedIndices.add(i);
    return selectedIndices;
  }

  public static <T> ArrayList<SelectableAdapterModel<T>> fromArray(T[] things) {
    ArrayList<SelectableAdapterModel<T>> newModels = new ArrayList<>();
    for (T thing : things) newModels.add(new SelectableAdapterModel<>(thing));
    return newModels;
  }

}

File: app/src/main/java/slak/gitforandroid/SelectableAdapterModel.kt
package slak.gitforandroid

import java.util.ArrayList

class SelectableAdapterModel<out T>(val thing: T) {
  var isSelected = false
    private set

  fun setSelectStatus(newStatus: Boolean) {
    isSelected = newStatus
  }

  companion object {
    fun <T> getSelectedModels(models: ArrayList<SelectableAdapterModel<T>>): ArrayList<Int> {
      val selectedIndices = models.indices.filterTo(ArrayList<Int>()) { models[it].isSelected }
      return selectedIndices
    }

    fun <T> fromArray(things: Array<T>): ArrayList<SelectableAdapterModel<T>> {
      val newModels = things.mapTo(ArrayList<SelectableAdapterModel<T>>()) {
        SelectableAdapterModel(it)
      }
      return newModels
    }
  }

}
